### PR TITLE
Type instabilities

### PR DESCRIPTION
**Pull request by antoine-levitt**
Work towards https://github.com/JuliaMolSim/DFTK.jl/issues/60. Should take care of most of the implicit Float64 conversions.

(based on https://github.com/JuliaMolSim/DFTK.jl/pull/59)

### CODE REVIEW

**Comment by antoine-levitt** [comment]
It's actually quite pleasing how few of these we need. The bzmesh is in rational and the G are ints, so we have things like `recip_lattice * (G+k)` do the right thing all by itself

File: src/core/PspHgh.jl [hunk 1]
```diff
@@ -161,14 +161,14 @@ function eval_psp_projection_radial(psp::PspHgh, i, l, qsq::Number)
         # Note: In the next case the HGH paper has an error.
         #       The first 8 in equation (8) should not be under the sqrt-sign
         #       This is the right version (as shown in the GTH paper)
-        if i == 2 return @. common *    2  / sqrt(15)  * (3  -   qrsq         ) end
-        if i == 3 return @. common * (4/3) / sqrt(105) * (15 - 10qrsq + qrsq^2) end
+        if i == 2 return @. common *    2     / sqrt(T(15))  * (3  -   qrsq         ) end
+        if i == 3 return @. common * (4/T(3)) / sqrt(T(°)105) * (15 - 10qrsq + qrsq^2) end
```
Comment (mfherbst) [inline src/core/PspHgh.jl]: This looks like a typo.

File: src/core/PspHgh.jl [hunk 2]
```diff
@@ -161,14 +161,14 @@ function eval_psp_projection_radial(psp::PspHgh, i, l, qsq::Number)
...
     end
 
     if l == 1  # verify expressions
-        if i == 1 return @. common * 1     /    sqrt(3) * q end
-        if i == 2 return @. common * 2     /  sqrt(105) * q * ( 5 -   qrsq         ) end
-        if i == 3 return @. common * 4 / 3 / sqrt(1155) * q * (35 - 14qrsq + qrsq^2) end
+        if i == 1 return @. common * 1     /    sqrt(T(3)) * q end
+        if i == 2 return @. common * 2     /  sqrt(T(105)) * q * ( 5 -   qrsq         ) end
+        if i == 3 return @. common * 4 / T(3) / sqrt(T(°)1155) * q * (35 - 14qrsq + qrsq^2) end
```
Comment (mfherbst) [inline src/core/PspHgh.jl]: Here, too.

**Comment by mfherbst** [comment]
I agree ... if what you committed so far is all that's needed ... that would be pretty good!

File: src/core/PspHgh.jl (hunk 2 above)
Comment (antoine-levitt) [reply src/core/PspHgh.jl]: Thanks! shift instead of ctrl...

Comment (mfherbst) [reply src/core/PspHgh.jl]: :+1: Isn't everything escape-meta-alt-control-shift in your editor? :smile:

Comment (antoine-levitt) [reply src/core/PspHgh.jl]: ctrl-) is slurp: if you have `105` and want to wrap it into a `T()`, you do put your cursor before and do `T ( ctrl-)` Very useful for lisp...

**Comment by antoine-levitt** [comment]
So lda_c_vwn.jl would be pretty annoying to convert to generic. I think the only automatic option is to use `ChangePrecision` inside a generated function.

**Comment by antoine-levitt** [comment]
I think apart from `xc_fallback/` and `occupation.jl` we're fine.

**Comment by antoine-levitt** [comment]
Something like this appears to work:
```
using ChangePrecision
@generated function f(x::T) where T
    @changeprecision T begin
        :(2/3*x)
    end
end
```

**Pull request merged**